### PR TITLE
many: implement the advisor backend, populate it from the store

### PR DESCRIPTION
**Pull request by chipaca**
This adds the backend to the advisor, and fills the backend from the store via the existing catalog refresh call.

It uses boltdb as the actual backend.

Based on #4438.

### CODE REVIEW

File: advisor/backend.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+	"github.com/snapcore/snapd/dirs"
+)
+
+var cmdBucketKey = []byte("Commands")
+
+type writer struct {
+	db *bolt.DB
+	tx *bolt.Tx
+	b  *bolt.Bucket
```
Comment (mvo5) [inline advisor/backend.go]: Maybe `b` for bucket is a bit too terse? When reading the code below I had to get back here to see what the `b` stands for.

File: advisor/backend.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+type CommandDB interface {
+	// AddSnap adds the entries for commands pointing to the given
+	// snap name to the commands database.
+	AddSnap(snapName string, commands []string) error
+	// Commit persist the changes, and closes the database.
+	Commit() error
+	// Rollback aborts the changes, and closes teh database.
```
Comment (mvo5) [inline advisor/backend.go]: Typo: `teh`

File: overlord/snapstate/catalogrefresh.go [hunk 3]
```diff
@@ -94,9 +97,24 @@ func refreshCatalogs(st *state.State, theStore StoreService) error {
...
 	}
 	defer namesFile.Cancel()
-	if err := theStore.WriteCatalogs(namesFile); err != nil {
+
+	cmdDB, err := newCmdDB()
+	if err != nil {
+		return err
+	}
+
+	defer cmdDB.Rollback()
```
Comment (mvo5) [inline overlord/snapstate/catalogrefresh.go]: Hm, I find this slightly confusing to have an unconditional rollback here? maybe a comment at least what is going on?

File: overlord/snapstate/catalogrefresh_test.go [hunk 4]
```diff
@@ -93,6 +97,15 @@ func (s *catalogRefreshTestSuite) TestCatalogRefresh(c *C) {
 	content, err = ioutil.ReadFile(dirs.SnapNamesFile)
 	c.Assert(err, IsNil)
 	c.Check(string(content), Equals, "pkg1\npkg2")
+
+	c.Check(osutil.FileExists(dirs.SnapCommandsDB), Equals, true)
+	dump, err := advisor.Dump()
```
Comment (mvo5) [inline overlord/snapstate/catalogrefresh_test.go]: nice!

File: overlord/snapstate/export_test.go [hunk 5]
```diff
@@ -146,6 +147,10 @@ func MockCatalogRefreshNextRefresh(cr *catalogRefresh, when time.Time) {
 	cr.nextCatalogRefresh = when
 }
 
+func MockCommandDB(constructor func() (advisor.CommandDB, error)) (restore func()) {
+	return nil
```
Comment (mvo5) [inline overlord/snapstate/export_test.go]: This looks a bit empty, no?

Comment (chipaca) [reply overlord/snapstate/export_test.go]: :innocent:

**Comment by chipaca** [comment]
FWIW while Update and View are convenient, I don't think they add that much to this -- and in particular note that the way I've implemented the populating of the database is possible because I'm not using those and doing the transaction manually (and doing the whole thing in a single function would need some further inversion of control which might not be too natural, at least as the code stands right now).

**Pull request merged**